source/utf8/checked.h

```diff
@@ -284,39 +299,142 @@ namespace utf8
           return (it == rhs.it);
       }
-      bool operator != (const iterator& rhs) const
+      bool operator != (const bidirectional_iterator& rhs) const
       {
           return !(operator == (rhs));
       }
-      iterator& operator ++ ()
+      bidirectional_iterator& operator ++ ()
       {
           utf8::next(it, range_end);
           return *this;
       }
-      iterator operator ++ (int)
+      bidirectional_iterator operator ++ (int)
       {
-          iterator temp = *this;
+          bidirectional_iterator temp = *this;
           utf8::next(it, range_end);
           return temp;
       }
-      iterator& operator -- ()
+      bidirectional_iterator& operator -- ()
       {
           utf8::prior(it, range_start);
           return *this;
       }
-      iterator operator -- (int)
+      bidirectional_iterator operator -- (int)
       {
-          iterator temp = *this;
+          bidirectional_iterator temp = *this;
           utf8::prior(it, range_start);
           return temp;
       }
-    }; // class iterator
+    }; // class bidirectional_iterator
+
+    template <typename octet_iterator>
+    class input_iterator : public std::iterator<std::input_iterator_tag, uint32_t> {
+      private:
+      octet_iterator it;
+      octet_iterator range_start;
+      octet_iterator range_end;
+      uint32_t cp{};
+      bool ok{};
+      void read() {
+			ok = it != range_end;
+			if(ok) {
+			   cp = utf8::next_impl(it, range_end);
+			}
+      }
+
+      public:
+      input_iterator () {}
+      explicit input_iterator (const octet_iterator& octet_it,
+                         const octet_iterator& rangestart,
+                         const octet_iterator& rangeend) :
+               it(octet_it), range_start(rangestart), range_end(rangeend)
+      {
+          read();
+      }
+      octet_iterator base () const { return it; }
+      uint32_t operator * () const
+      {
+          if(!ok) {
+				 throw std::runtime_error("no such element");
+          }
+          return cp;
+      }
+
+      bool operator == (const input_iterator& rhs) const
+      {
+          if (range_start != rhs.range_start || range_end != rhs.range_end)
+              throw std::logic_error("Comparing utf-8 iterators defined with different ranges");
+          return ok == rhs.ok && (!ok || it == rhs.it);
+      }
+      bool operator != (const input_iterator& rhs) const
+      {
+          return !(operator == (rhs));
+      }
+      input_iterator& operator ++ ()
+      {
+          ++it;
+          read();
+          return *this;
+      }
+      input_iterator operator ++ (int)
+      {
+          input_iterator temp = *this;
+          ++it;
+          read();
+          return temp;
+      }
+    }; // class input_iterator
+
+    template <typename octet_iterator>
+    struct get_iterator_class {
+    private:
+		static input_iterator<octet_iterator> get(std::input_iterator_tag);
```

Comment:
Maybe add a static_assert here that it is indeed an octet iterator? Like `sizeof(decltype(*std::declval<octet_iterator>())) == 8`

Reply:
Interesting but I don't think I will add it. Someone else can add it.